Stateful Models Documentation Update (#22410)
openvinotoolkit/openvino: docs/articles_en/openvino_workflow/running_inference_with_openvino/stateful_models_intro.rst

.. {#openvino_docs_OV_UG_stateful_models_intro}

Stateful models and State API
==============================

.. toctree::
    :maxdepth: 1
    :hidden:

    openvino_docs_OV_UG_ways_to_get_stateful_model

What is Stateful Model?
#######################

> Stateful model is a model which stores the specified data between and during model inferences in internal memory space usually called State or Variable.
> In contrast to usual **stateless** model, which purify all the calculated data after each inference call, **stateful**
> model preserve tensors saved in States.

Reviewer: ```suggestion
Stateful model is a model which implicitly keeps data from one inference call to the next inference call. Data is kept in internal runtime memory space usually called State or Variable.
In contrast to usual **stateless** model, which return all produced data as model outputs, **stateful**
model preserve part of the tensors saved in States without exposing them as model outputs.

The purpose of stateful models is to natively address a sequence processing like in text generation tasks when one model inference produce a single output token, and it is required to perform multiple inference calls to generate a complete output sentence. Hidden state data from previous inference should be passed to the next inference as a context. Usually the contextual data is not required to be accessed in the user application and should be just passed through to the next inference call manually using model API. Stateful models simplifies programming of this scenario and unlocks additional performance potential of OpenVINO runtime.
```

OpenVINO Stateful Model Representation
######################################

OpenVINO contains ReadValue/Assign operations to make a model Stateful.
Each pair of ReadValue/Assign operates with State, known also as Variable,
which is an internal memory buffer to store tensor data during and between model inference calls.
ReadValue reads data tensor from State and returns it as output, Assign accepts data tensor as input and writes data to State

> to save data for the next inference call.

Reviewer: Move the first paragraph to a more low-level explanation below. Rename this section to Benefits of Stateful Models and list those 3 items below without mentioning ReadValue and Assign ops. Introduce low-level stuff gradually later. The beginning of the article shouldn't touch op-level specifics, first explain purpose, benefits and ways to obtain stateful models, then move to API specifics and inherent limitations that helps runtime to unlock that benefits that were promised in the beginning in comparison to regular Parameter/Results.

These entities work as a coherent mechanism and help to achieve the next goals:

1. Speed up execution of Model
    Data in State is stored in the optimized form for OpenVINO plugins, which helps to execute model effectively.
    But often requesting data from State might reduce the expected performance gains and even lead to losses,

>     so, it is expected that State mechanism will be used when data stored in State is not accessed frequently.

Reviewer: We need to be more careful with performance promises. It really depends on the model architecture (patterns used between ReadValue and Assign ops in the graph). Please be more specific and state that performance improvements are expected specifically for large language models produced by Optimum-Intel path and for other models it may vary.

Reviewer: I mean even state is not accessed and developer plays by the rules, it may impact performance for some architectures. It is hard to explain when it happens, it is better to point to text generation with model produced by Optimum-Intel where we are definitely expect performance gain.

2. Simplify user code
    Typical scenarios as copying data from model's outputs to inputs in user code can be replaced with ReadValue/Assign operations.
    OpenVINO will manage this copying internally.

3. Specific scenarios
    Several use cases require processing of data sequences. When length of a sequence is known and small enough,

>     we can process it with RNN like models that contain a cycle inside. But in some cases, like online speech recognition of time series

Reviewer 2: ```suggestion
    we can process it with RNN like models that contain a cycle inside. But in some cases, like online speech recognition or time series
```

    forecasting, length of data sequence is unknown. Then data can be divided in small portions and processed step-by-step. But dependency
    between data portions should be addressed. For that, models save some data between inferences - state. When one dependent sequence is over,
    state should be reset to initial value and new sequence can be started.

OpenVINO has a special API to simplify work with Stateful models. State is automatically saved between inferences,

> and there is a way to reset state when needed. You can also read state or set it to some new value between inferences.

Reviewer: Please rephrase it to be more fluent: we have already state that we have a dedicated API for that. And it looks to be a duplication. Provide more details (e.g. InferRequest class reference to identify where this API exists).

Stateful Model Example
######################

.. image:: _static/images/stateful_model_example.png
   :align: center

> The left side of the picture shows the usual inputs and outputs to the model: Parameter/Result operations. They have no connection with each other and in order to copy data from output to input users need to put extra effort writing and maintaining additional code.

Reviewer 2: ```suggestion
The left side of the picture shows the usual inputs and outputs to the model: Parameter/Result operations. There's no direct connection from Result to Parameter and in order to copy data from output to input users need to put extra effort writing and maintaining additional code.
```
In fact there are always two connections, just indirect: Parameter->Result through model and Result->Parameter through code

.. image:: _static/images/stateful_model_init_subgraph.png
   :align: center

In some cases, users need to set an initial value for State, or it may be necessary to reset the value of State at a certain inference to the initial value. For such situations, an initializing subgraph for the ReadValue operation and a special "reset" method are provided.

You can find more details on these operations in [ReadValue specification](../ops/infrastructure/ReadValue_6.md) and
[Assign specification](../ops/infrastructure/Assign_6.md).

How to get OpenVINO Model with States
#########################################

* [Apply LowLatency2 transformation.](./ways_to_get_stateful_model.md#)
   If a model contains a loop that runs over some sequence of input data,
   the LowLatency2 transformation can be applied to get model with states.
   Note: there are some [specific limitations]() to use the transformation.

* [Apply MakeStateful transformation.](./ways_to_get_stateful_model.md)
   If after conversion from original model to OpenVINO representation, the resulting model contains Parameter and Result operations,

>    which pairwise have the same shape and element type, the MakeStateful transformation can be applied to get model with states.

Reviewer: Promote Optimum-Intel as the main source of stateful models. Refer to Gen AI article in this documentation. Position LowLatency2 and MakeStateful as more low-level/advanced technique for developers who really know what they want.

Author: done
